src/flag_gems/ops/scatter.py

```diff
@@ -1,216 +1,300 @@
+    code.newline()
+
+    # the autotune function
+    code.writeline("def cfggen():")
+    with code.indent():
+        code.writeline("block_m = [1, 2, 4, 8]")
+        code.writeline("block_n = [256, 512, 1024, 2048]")
+        code.writeline("configs = [")
+        with code.indent():
+            code.writeline('triton.Config({"BLOCK_M": m, "BLOCK_N": n}, num_warps=4)')
+            code.writeline("for m in block_m")
+            code.writeline("for n in block_n")
+        code.writeline("]")
+        code.writeline("return configs")
+
+    code.newline()
+    code.newline()
+
+    # the decorators
+    code.writeline("@libentry()")
+    code.writeline('@triton.autotune(configs=cfggen(), key=["M", "N"])')
+    code.writeline("@triton.jit")
+
+    # signature
+    code.writeline(f"def {kernel_name}(")
+    function_ns = NameSpace()
+    with code.indent():
+        if rank > 0:
+            code.writeline("src_strided,")
+            function_ns.create_name("src_strided")
+            code.writeline("index,")
+            function_ns.create_name("index")
+            code.writeline("inp,")
+            function_ns.create_name("inp")
+            code.writeline("out,")
+            function_ns.create_name("out")
+
+            for i in range(rank):
+                function_ns.create_name(f"inp_stride_{i}")
+            stride_args = ", ".join(f"inp_stride_{i}: int" for i in range(rank))
+            code.writeline(f"{stride_args}, # stride for inp")
+
+            for i in range(rank):
+                function_ns.create_name(f"index_stride_{i}")
+            stride_args = ", ".join(f"index_stride_{i}: int" for i in range(rank))
+            code.writeline(f"{stride_args}, # stride for index")
+
+            for i in range(rank):
+                function_ns.create_name(f"index_shape_{i}")
+            shape_args = ", ".join(f"index_shape_{i}: int" for i in range(rank))
+            code.writeline(f"{shape_args}, # shape for index")
+
+            code.writeline("dim,")
+            code.writeline("stride_dim,")
+            code.writeline("M,")
+            code.writeline("N,")
+            # reduce options
+            code.writeline("IS_ADD: tl.constexpr,")
+            code.writeline("IS_MUL: tl.constexpr,")
+            code.writeline("BLOCK_M: tl.constexpr,")
+            code.writeline("BLOCK_N: tl.constexpr,")
+
+    code.writeline("):")
+
+    # Kernel Code
+    with code.indent():
+        code.writeline("pid_x = tl.program_id(0)")
+        code.writeline("pid_y = tl.program_id(1)")
+        code.writeline(
+            "rows_offsets = pid_x * BLOCK_M + tl.arange(0, BLOCK_M)[:, None]"
+        )
+        code.writeline(
+            "cols_offsets = pid_y * BLOCK_N + tl.arange(0, BLOCK_N)[None, :]"
+        )
+        code.writeline("rows_mask = rows_offsets < M")
+        code.writeline("cols_mask = cols_offsets < N")
+
+        code.writeline("offsets = rows_offsets * N + cols_offsets")
```

Comment:
I think rows_offset * N could overflow for large N. Is it safe not casting to int64 here?

Reply:
While current tests haven't shown related errors, I understand your concerns and it does make sense. I will do upcasting for index calculations.